flink-end-to-end-tests/flink-streaming-kafka-test-base/src/main/java/org/apache/flink/streaming/kafka/test/base/KafkaExampleUtil.java

```diff
@@ -45,6 +45,7 @@ public static StreamExecutionEnvironment prepareExecutionEnv(ParameterTool param
 		env.getConfig().disableSysoutLogging();
 		env.getConfig().setRestartStrategy(RestartStrategies.fixedDelayRestart(4, 10000));
 		env.enableCheckpointing(5000); // create a checkpoint every 5 seconds
+		env.getCheckpointConfig().setTolerableCheckpointFailureNumber(Integer.MAX_VALUE);
```

Comment:
This one I can make sure, the Kafka end to end test script checked the log file is not empty and the log reported: the count is execeed the tolerance number.

Reply:
Ok, then the question is through which failure reason, and if there was a higher tolerance for that reason before. Could it be DECLINED was tolerated before, e.g. when tasks where not yet ready?

Reply:
I see many waiting JobMaster print, but not very sure. I think it could be declined before this PR. It was OK is because the old tolerance mechanism exists TM end. But, now, it would trigger restart by `CheckpointFailureManager`.

Reply:
After a quick comparison back with master and 1.7 I think there is at least a problem with the `DECLINED` case. It was treated similar to e.g. subsumed and was never leading to a job failure. That also makes sense because this cause is just existing because the JM can already start triggering checkpoints before all tasks are running. This is something that is currently (unfortunately) to expect and that should not lead to a failover because it can happen regularly in the beginning of a job. Wdyt? If you agree let's also double-check the other cases one more time.

Reply:
I agree. So, to see more details do you agree I remove the other settings about tolerance number. Let Travis replay all possible situations?

Reply:
Yes, in my opinion the test should pass without any change to that option, or else we changed some behaviour. And I am pretty sure that at least this case is different in the current PR than from what happened before.

Reply:
OK, I will follow the steps I listed lie. Let's analyze them case by case.

Reply:
Maybe also one more hint, whenever you created a stub for the failure manager in the tests, the fail callback was an empty implementation. We should be sure that that is either ok, or maybe better have something that throws an exception to indicate that this mechanism was maybe accidentally triggered by a breaking change?

Reply:
OK, I will try to refactor the test cases which contain the failure manager.

Reply:
Have added a new test case to verify the exception behavior.